src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs

```diff
@@ -195,5 +195,233 @@ public static void InvokeFunctionAcceptingArrayTypes(Function objectPrototype, s
+                });");
+
+            var promise = (Task<object>)factory.Call();
+
+            var ex = await Assert.ThrowsAsync<JSException>(async () => await promise);
+            Assert.Equal("Error: fail", ex.Message);
+        }
+
+
+        [ActiveIssue("not implemented")]
+        [Fact]
+        public static void RoundtripPromise()
+        {
+            var factory = new Function(@"
+                var dummy=new Promise((resolve, reject) => {});
+                return {
+                    dummy:dummy,
+                    check:(promise)=>{
+                        console.log(JSON.stringify(promise));
+                        return promise===dummy ? 1:0;
+                    }
+                }");
+
+            var obj = (JSObject)factory.Call();
+            var dummy = obj.GetObjectProperty("dummy");
+            Assert.IsType<Task<object>>(dummy);
+            var check = obj.Invoke("check", dummy);
+            Assert.Equal(1, check);
+        }
+
+
+        [Fact]
+        public static async Task ResolveTask()
+        {
+            var tcs = new TaskCompletionSource<int>();
+            var factory = new Function("task", "callback", @"
+                return task.then((data)=>{
+                    callback(data);
+                })
+            ");
+
+            int check = 0;
+            Action<int> callback = (int data) =>
+            {
+                check = data;
+            };
+            var promise = (Task<object>)factory.Call(null, tcs.Task, callback);
+            Assert.Equal(0, check);
+            tcs.SetResult(1);
+            Assert.Equal(0, check);
+            await promise;
+            Assert.Equal(1, check);
+        }
+
+        [Fact]
+        public static async Task RejectTask()
+        {
+            var tcs = new TaskCompletionSource<int>();
+            var factory = new Function("task", "callback", @"
+                return task.catch((reason)=>{
+                    callback(reason);
+                })
+            ");
+
+            string check = null;
+            Action<string> callback = (string data) =>
+            {
+                check = data;
+            };
+            var promise = (Task<object>)factory.Call(null, tcs.Task, callback);
+            Assert.Null(check);
+            tcs.SetException(new Exception("test"));
+            Assert.Null(check);
+            await promise;
+            Assert.Contains("System.Exception: test", check);
+        }
+
+        [Fact]
+        public static void NullTask()
+        {
+            var tcs = new TaskCompletionSource<int>();
+            var factory = new Function("task", "callback", @"
+                callback(task);
+            ");
+
+            Task check = null;
+            Action<Task> callback = (Task data) =>
+            {
+                check = data;
+            };
+            Assert.Null(check);
+            factory.Call(null, null, callback);
+            Assert.Null(check);
```

Comment:
I don't think this is verifying what you intend for it to verify - if data is ```null``` when callback is invoked, it will overwrite the ```null``` in ```check``` with ```null```. You probably want a bool that records whether callback was invoked, in addition to verifying the value.